### PR TITLE
feat :async test client prototype

### PR DESCRIPTION
**Pull request by kareemmahlees**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [ ] Have you updated the prose documentation?
- [ ] Have you updated the reference documentation?

related issue #692
# Modifications
- created an async client class, transport, and a life span handler to `starlite\testing\async_test_client`
- the client until now only has the `request` method
- created a test case that makes sure the client is running 
- renamed `test_client` to `sync_test_client` and `async_test_client`

All the added code is basically the same as the `sync_test_client` but with modifications to suit the async/await 

I would like to hear your feedback on this so far.

Note : i tried to fix flake8 pre-commit error with multible ways but i couldn't reach to a working solution

### CODE REVIEW

File: starlite/testing/async_test_client/transport.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                request_event["body"] = body if isinstance(body, bytes) else body.encode("utf-8")
+            return request_event
+
+        return receive
+
+    @staticmethod
+    def create_send(request: "Request", context: SendReceiveContext) -> "Send":
+        async def send(message: "Message") -> None:
+            if message["type"] == "http.response.start":
+                if context["response_started"]:
```
Comment (kareemmahlees) [inline starlite/testing/async_test_client/transport.py:82]: this is the statement where flake8 has problems, even after changing the `assert statement` to `if statement` and running the flake8 pre-commit hook locally ( which passes ) it still has some issues here in the ci
````diff
- lkjaslkdjfklasdf
+ klajslkdjfa
````

File: starlite/testing/async_test_client/transport.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+                    raise AssertionError('Received multiple "http.response.start" messages.')
+                context["raw_kwargs"]["status_code"] = message["status"]
+                context["raw_kwargs"]["headers"] = [
+                    (k.decode("utf-8"), v.decode("utf-8")) for k, v in message.get("headers", [])
+                ]
+                context["response_started"] = True
+            elif message["type"] == "http.response.body":
+                if not context["response_started"]:
```
Comment (kareemmahlees) [inline starlite/testing/async_test_client/transport.py:90]: this is the statement where flake8 has problems, even after changing the `assert statement` to `if statement` and running the flake8 pre-commit hook locally ( which passes ) it still has some issues here in the ci
````diff
- assert context["response_started"] , 'Received "http.response.body" without "http.response.start".'
+ if not context["response_started"]:
+ raise AssertionError('Received "http.response.body" without "http.response.start".')
````

File: starlite/testing/async_test_client/transport.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                    raise AssertionError('Received "http.response.body" without "http.response.start".')
+                if context["response_complete"].is_set():
```
Comment (kareemmahlees) [inline starlite/testing/async_test_client/transport.py:92]: this is the statement where flake8 has problems, even after changing the `assert statement` to `if statement` and running the flake8 pre-commit hook locally ( which passes ) it still has some issues here in the ci
````diff
- assert not context["response_started"] , 'Received "http.response.body" after response completed.'
+ if context["response_complete"].is_set():
+ raise AssertionError('Received "http.response.body" after response completed.')
````

File: starlite/testing/async_test_client/transport.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        except BaseException as exc:  # pragma: no cover # pylint: disable=W0703
+            if self.raise_server_exceptions:
+                raise exc
+            return Response(
+                status_code=HTTP_500_INTERNAL_SERVER_ERROR, headers=[], stream=ByteStream(b""), request=request
+            )
+        else:
+            if not context["response_started"]:  # pragma: no cover
+                if self.raise_server_exceptions:  # noqa: SIM102
+                    if not context["response_started"]:
```
Comment (kareemmahlees) [inline starlite/testing/async_test_client/transport.py:177]: this is the statement where flake8 has problems, even after changing the `assert statement` to `if statement` and running the flake8 pre-commit hook locally ( which passes ) it still has some issues here in the ci
````diff
- assert context["response_started"] , "TestClient did not receive any response."
+  if not context["response_started"]:
+ raise AssertionError("TestClient did not receive any response.")

File: starlite/testing/async_test_client/client.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+T = TypeVar("T", bound=ASGIApp)
+
+
+def fake_http_send_message(headers: MutableScopeHeaders) -> HTTPResponseStartEvent:
+    headers.setdefault("content-type", "application/text")
+    return HTTPResponseStartEvent(type="http.response.start", status=200, headers=headers.headers)
+
+
+def fake_asgi_connection(app: ASGIApp, cookies: Dict[str, str]) -> ASGIConnection[Any, Any, Any]:
```
Comment (Goldziher) [inline starlite/testing/async_test_client/client.py:56]: this is duplicated with the "sync_client". Please reduce duplications.

File: starlite/testing/async_test_client/client.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        URLTypes,
+    )
+
+    from starlite.middleware.session.base import BaseBackendConfig, BaseSessionBackend
+
+
+T = TypeVar("T", bound=ASGIApp)
+
+
+def fake_http_send_message(headers: MutableScopeHeaders) -> HTTPResponseStartEvent:
```
Comment (Goldziher) [inline starlite/testing/async_test_client/client.py:51]: duplicated code

File: starlite/testing/async_test_client/client.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        "path_params": {},
+        "route_handler": None,
+        "_cookies": cookies,
+    }
+    return ASGIConnection[Any, Any, Any](
+        scope=scope,  # type: ignore[arg-type]
+    )
+
+
+class AsyncTestClient(AsyncClient, Generic[T]):
```
Comment (Goldziher) [inline starlite/testing/async_test_client/client.py:80]: So, there is a lot of duplication happening here. Can't we inherit from the sync test client and simply override `exit_stack`?

File: starlite/testing/async_test_client/transport.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+                if not more_body:
+                    context["raw_kwargs"]["stream"].seek(0)
+                    context["response_complete"].set()
+            elif message["type"] == "http.response.template":  # type: ignore[comparison-overlap] # pragma: no cover
+                context["template"] = message["template"]  # type: ignore[unreachable]
+                context["context"] = message["context"]
+
+        return send
+
+    def parse_request(self, request: "Request") -> Dict[str, Any]:
```
Comment (Goldziher) [inline starlite/testing/async_test_client/transport.py:107]: again, there is a lot of duplication happening here. Cant this be refactored to use inheritence?

**Review by Goldziher — COMMENTED** [review]
Thanks for contributing @kareemmahlees! I left several comments. 

I would regard this PR as in a PoC stage at present. Please refactor the code to use inheritence and reduce duplications. 

The other important factor is to ensure that editor completion is in place (this is why all the get/post/put etc. methods are implemented for the sync client). 

Finally, the tests need to be substantially increased. I wouldn't consider the code tested at all with the given state of the single test in place.

File: starlite/testing/async_test_client/client.py (hunk 5 above)
Comment (kareemmahlees) [reply starlite/testing/async_test_client/client.py]: this solves it [149a817](https://github.com/starlite-api/starlite/pull/980/commits/149a817b5cd84819582725c0ada2a9bbdd5d079d)

File: starlite/testing/async_test_client/client.py (hunk 7 above)
Comment (kareemmahlees) [reply starlite/testing/async_test_client/client.py]: this solves it [149a817](https://github.com/starlite-api/starlite/pull/980/commits/149a817b5cd84819582725c0ada2a9bbdd5d079d)

File: starlite/testing/async_test_client/client.py (hunk 6 above)
Comment (kareemmahlees) [reply starlite/testing/async_test_client/client.py]: this solves it [149a817](https://github.com/starlite-api/starlite/pull/980/commits/149a817b5cd84819582725c0ada2a9bbdd5d079d)

File: starlite/testing/async_test_client/transport.py (hunk 8 above)
Comment (kareemmahlees) [reply starlite/testing/async_test_client/transport.py]: this solves it [149a817](https://github.com/starlite-api/starlite/pull/980/commits/149a817b5cd84819582725c0ada2a9bbdd5d079d)

**Comment by kareemmahlees** [comment]
@Goldziher could you check the current changes!

**Comment by Goldziher** [comment]
> @Goldziher could you check the current changes!

sure. Will take me some time to go throufh this, but will do this. Is it ready for review? if it is, change it from draft to PR mode. 

Also, please rebase on main and resolve the conflicts.

**Pull request closed without merging**